single-line-diagram/single-line-diagram-core/src/main/java/com/powsybl/sld/layout/zonebygrid/MatrixZoneLayoutModel.java

```diff
@@ -0,0 +1,210 @@
+import com.powsybl.sld.model.graphs.*;
+
+import java.util.*;
+
+/**
+ * @author Thomas Adam <tadam at neverhack.com>
+ */
+public class MatrixZoneLayoutModel {
+
+    private final Map<String, MatrixCell> cellsById = new HashMap<>();
+    private final List<MatrixCell> emptyCells = new ArrayList<>();
+
+    private int matrixNbRow = 0;
+
+    private int matrixNbCol = 0;
+
+    private int matrixCellHeight = -1;
+
+    private int matrixCellWidth = -1;
+
+    private final int snakelineHallwayWidth;
+
+    private Grid pathFinderGrid;
+
+    public MatrixZoneLayoutModel(int hallway) {
+        this.snakelineHallwayWidth = hallway;
+    }
+
+    public void addGraph(SubstationGraph graph, int col, int row) {
+        if (graph != null) {
+            cellsById.put(graph.getId(), new MatrixCell(graph, col, row));
+            matrixCellHeight = Math.max(matrixCellHeight, (int) graph.getHeight());
+            matrixCellWidth = Math.max(matrixCellWidth, (int) graph.getWidth());
+        } else {
+            emptyCells.add(new MatrixCell(null, col, row));
+        }
+        matrixNbRow = Math.max(matrixNbRow, row + 1);
+        matrixNbCol = Math.max(matrixNbCol, col + 1);
+    }
+
+    public int getSnakelineHallwayWidth() {
+        return snakelineHallwayWidth;
+    }
+
+    public int getMatrixCellWidth() {
+        return matrixCellWidth;
+    }
+
+    public int getMatrixCellHeight() {
+        return matrixCellHeight;
+    }
+
+    public int getX(String id) {
+        MatrixCell cell = cellsById.get(id);
+        return getX(cell.col());
+    }
+
+    public int getX(int col) {
+        return ((col + 1) * snakelineHallwayWidth) + (col * matrixCellWidth);
+    }
+
+    public int getY(String id) {
+        return getY(id, Direction.TOP);
+    }
+
+    public int getY(int row) {
+        return getY(row, Direction.TOP);
+    }
+
+    public int getY(int row, Direction direction) {
+        return (row + 1) * snakelineHallwayWidth + (row + (direction == Direction.TOP ? 0 : 1)) * matrixCellHeight;
+    }
+
+    public int getY(String id, Direction direction) {
+        MatrixCell cell = cellsById.get(id);
+        int row = cell.row();
+        return getY(row, direction);
+    }
+
+    public List<Point> buildSnakeline(PathFinder pathfinder,
+                                      String ss1Id, Point p1, Direction d1,
+                                      String ss2Id, Point p2, Direction d2) {
+        insertFreePathInSubstation(ss1Id, p1, d1, ss2Id, p2, d2);
+
+        // Use path finding algo
+        List<Point> snakeLine = pathfinder.toSnakeLine(pathfinder.findShortestPath(pathFinderGrid,
+                (int) p1.getX(), (int) p1.getY(),
+                (int) p2.getX(), (int) p2.getY(),
+                true));
+
+        // Only for debug
+        pathFinderGrid.dumpToFile("out.txt");
+
+        return snakeLine;
+    }
+
+    private void insertFreePathInSubstation(String id1, Point p1, Direction d1,
```

Comment:
This function contains duplicate code for each of the points P1 and P2. I have the impression that P1 and P2 do not interact with each other in `insertFreePathInSubstation`. Maybe we could have `insertFreePathInSubstation(ssId, p, d)` and call the function twice on `p1` and `p2`?

Reply:
In deed